### PR TITLE
Add generic dao filtering action based on comparing object properties

### PR DESCRIPTION
**Pull request by nanoMichal**
The goal here is to return a dao containing objects that match the provided "new object's" properties. Properties are specified in the properties array, and what to do with the dao is left up to the user who extends this model.

### CODE REVIEW

**Comment by nanoMichal** [comment]
The goal here is to return a dao containing objects that match the provided "new object's" properties. Properties are specified in the properties array, and what to do with the dao is left up to the user who extends this model.

File: src/foam/nanos/ruler/action/AbstractCheckDAOforMatching.js [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  package: 'foam.nanos.ruler.action',
+  name: 'AbstractCheckDAOforMatching',
+  abstract: true,
+
+  documentation: `do a dao call to return a predicated dao of objects matching the properties specified by an input. `,
+
+  implements: ['foam.nanos.ruler.RuleAction'],
+
+  javaImports: [
+    'foam.core.FObject',
```
Comment (kgrgreer) [inline src/foam/nanos/ruler/action/AbstractCheckDAOforMatching.js:18]: sort your imports

File: src/foam/nanos/ruler/action/AbstractCheckDAOforMatching.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    'foam.dao.DAO',
+    'foam.nanos.logger.Logger',
+    'foam.core.ContextAgent',
+    'foam.mlang.predicate.Predicate',
+    'static foam.mlang.MLang.*'
+  ],
+
+  properties: [
+    {
+      class: 'Array',
```
Comment (kgrgreer) [inline src/foam/nanos/ruler/action/AbstractCheckDAOforMatching.js]: class: 'StringArray', which is a separate type.

File: src/foam/nanos/ruler/action/AbstractCheckDAOforMatching.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+      name: 'properties',
+      javaFactory: 'return new String[0];'
+    },
+    {
+      class: 'String',
+      name: 'daoName'
+    },
+    {
+      class: 'Boolean',
+      name: 'useAgent'
```
Comment (kgrgreer) [inline src/foam/nanos/ruler/action/AbstractCheckDAOforMatching.js]: Add documentation:

File: src/foam/nanos/ruler/action/AbstractCheckDAOforMatching.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+            public void execute(X x) {
+              DAO myDAO = ((DAO) x.get( getDaoName() )).where(myPredicate);
+              doThis(x, nu, myDAO);
+            }
+          },"put the class name of this rule");
+        }
+      `
+    },
+    {
+      name: 'doThis',
```
Comment (kgrgreer) [inline src/foam/nanos/ruler/action/AbstractCheckDAOforMatching.js]: How about naming something like: 'execute' or 'cmd'?

File: src/foam/nanos/ruler/action/AbstractCheckDAOforMatching.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+      args: [
+        { name: 'x', type: 'Context' },
+        { name: 'nu', type: 'foam.core.FObject' },
+        { name: 'dao', type: 'foam.dao.DAO' },
+      ],
+      javaCode: `
+       // Your Code Here.
+      `
+    },
+
```
Comment (kgrgreer) [inline src/foam/nanos/ruler/action/AbstractCheckDAOforMatching.js]: Change to:
// Template method, add code in sub-class

File: src/foam/nanos/ruler/action/AbstractCheckDAOforMatching.js [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+            public void execute(X x) {
+              DAO myDAO = ((DAO) x.get( getDaoName() )).where(myPredicate);
+              execute(x, nu, myDAO);
+            }
+          },"DAO find Operation");
+        }
+      `
+    },
+    {
+      name: 'execute',
```
Comment (nanoMichal) [inline src/foam/nanos/ruler/action/AbstractCheckDAOforMatching.js]:
```suggestion
      name: 'cmd',
```

File: src/foam/nanos/ruler/action/AbstractCheckDAOforMatching.js [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        if ( getReadOnly()) {
+          DAO myDAO = ((DAO) x.get( getDaoName() )).where(myPredicate);
+          execute(x, nu, myDAO);
+        }
+        else {
+          agency.submit(x, new ContextAgent() {
+            @Override
+            public void execute(X x) {
+              DAO myDAO = ((DAO) x.get( getDaoName() )).where(myPredicate);
+              execute(x, nu, myDAO);
```
Comment (nanoMichal) [inline src/foam/nanos/ruler/action/AbstractCheckDAOforMatching.js]:
```suggestion
              cmd(x, nu, myDAO);
```

File: src/foam/nanos/ruler/action/AbstractCheckDAOforMatching.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        Predicate [] propertiesToCheck = new Predicate [getProperties().length] ;
+
+        for ( int i = 0; i < getProperties().length; i++ ) {
+          propertiesToCheck[i] = EQ( nu.getClassInfo().getAxiomByName(getProperties()[i]), nu.getProperty(getProperties()[i]) );
+        }
+        Predicate myPredicate = AND(propertiesToCheck);
+
+        if ( getReadOnly()) {
+          DAO myDAO = ((DAO) x.get( getDaoName() )).where(myPredicate);
+          execute(x, nu, myDAO);
```
Comment (nanoMichal) [inline src/foam/nanos/ruler/action/AbstractCheckDAOforMatching.js]:
```suggestion
          cmd(x, nu, myDAO);
```

**Pull request merged**